x-pack/plugin/blob-cache/src/main/java/org/elasticsearch/blobcache/shared/SharedBlobCacheService.java

```diff
@@ -421,6 +437,62 @@ public Entry<CacheFileRegion> get(KeyType cacheKey, long fileLength, int region)
         return entry;
     }
 
+    /**
+     * Fetch and cache the full blob for the given cache entry from the remote repository if there
+     * are enough free pages in the cache to do so.
+     *
+     * This method returns as soon as the download tasks are instantiated, but the tasks themselves
+     * are run on the bulk executor.
+     *
+     * If an exception is thrown from the writer then the cache entry being downloaded is freed
+     * and unlinked
+     *
+     * @param cacheKey  the key to fetch data for
+     * @param length    the length of the blob to fetch
+     * @param writer    a writer that handles writing of newly downloaded data to the shared cache
+     *
+     * @return {@code true} if there were enough free pages to start downloading
+     */
+    public boolean maybeFetchFullEntry(KeyType cacheKey, long length, RangeMissingHandler writer) {
+        int finalRegion = getEndingRegion(length);
+        if (freeRegionCount() < finalRegion) {
+            // Not enough room to download a full file without evicting existing data, so abort
```

Comment:
This won't be totally accurate under load, but I don't think that matters - if the cache is nearly full then we will see eviction pressure under normal search load whether or not we're doing a prefetch.

Reply:
++ makes sense to me, the cache behaviour totally degrades in that case anyway, this wouldn't be where we would have to start fixing things :)